pkg/distributor/streamsharder.go

```diff
@@ -0,0 +1,45 @@
+package distributor
+
+import (
+	"sync"
+
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+type streamSharder struct {
+	mu      sync.Mutex
+	streams map[string]int
+}
+
+func NewStreamSharder() StreamSharder {
+	return &streamSharder{
+		streams: make(map[string]int),
+	}
+}
+
+func (s *streamSharder) ShardsFor(stream logproto.Stream) (int, bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	shards := s.streams[stream.Labels]
+	if shards > 0 {
+		return shards, true
+	}
+
+	return 0, false
+}
+
+func (s *streamSharder) IncreaseShardsFor(stream logproto.Stream) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	shards := s.streams[stream.Labels]
+	s.streams[stream.Labels] = max(shards*2, 2)
```

Comment:
I think this needs some explanation.

Reply:
Right, we can add a docstring for that. But explaining here anyway: this is to make sure the first sharding will result in at least 2 shards. Since it is the first sharding, shards would be 0, so shards * 2 = 0 which wouldn't work. This is basically to avoid doing:
```
if (shards) == 0 {
  shards = 1
}
s.streams[stream.Labels] = shards*2
```

Reply:
I pushed a commit handling it, lmk what you think.

Reply:
Thanks. That clears it up.